From 04e731901f212c51c5926c536bd3129125131ef7 Mon Sep 17 00:00:00 2001
From: Ilan Peer <ilan.peer@intel.com>
Date: Fri, 22 Dec 2023 11:04:18 +0200
Subject: [PATCH] WLANRTSYS-82700 AP: Add configuration options for 6 GHz TPE
 Tx power

Add configuration options for setting the Tx Power value
in the Transmit Power Envelope for 6 GHz:

- The Tx power value for default client where the transmit
  power interpretation is "Regulatory Client EIRP PSD"
- The Tx power value for subordinate client where the transmit
  power interpretation is "Regulatory Client EIRP PSD"

Signed-off-by: Ilan Peer <ilan.peer@intel.com>
---
 hostapd/config_file.c |  4 ++++
 hostapd/hostapd.conf  |  7 +++++++
 src/ap/ap_config.c    |  2 ++
 src/ap/ap_config.h    |  4 ++++
 src/ap/ieee802_11.c   | 11 +++++++++--
 5 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index bb0a8ac9c..38d983865 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -4450,6 +4450,10 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 				   line);
 			return 1;
 		}
+	} else if (os_strcmp(buf, "reg_def_cli_eirp_psd") == 0) {
+		conf->reg_def_cli_eirp_psd = atoi(pos);
+	} else if (os_strcmp(buf, "reg_sub_cli_eirp_psd") == 0) {
+		conf->reg_sub_cli_eirp_psd = atoi(pos);
 	// ToDo: remove this and keep opensource configuration for SRPS IE
 	} else if (os_strcmp(buf, "he_spatial_reuse_ie_present_in_beacon") == 0) {
 		conf->he_spatial_reuse_ie_present_in_beacon = atoi(pos);
diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index 0b65b0af2..913bc39b8 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1049,6 +1049,13 @@ wmm_ac_vo_acm=0
 # See IEEE P802.11-REVme/D4.0, Table E-12 (Regulatory Info subfield encoding)
 # for more details.
 #he_6ghz_reg_pwr_type=0
+#
+# 6 GHz Maximum Tx Power used in Transmit Power Envelope elements, where the
+# "Transmit Power Interpretation" is set to "Regulatory client EIRP PSD".
+# For Maximum Transmit Power Category subfield encoding set to default (0):
+#reg_def_cli_eirp_psd=-1
+# For Maximum Transmit Power Category subfield encoding set to subordinate (1):
+#reg_sub_cli_eirp_psd=-1
 
 # Unsolicited broadcast Probe Response transmission settings
 # This is for the 6 GHz band only. If the interval is set to a non-zero value,
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index 4e987aa01..ae55d91b2 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -356,6 +356,8 @@ struct hostapd_config * hostapd_config_defaults(void)
 	conf->he_6ghz_rx_ant_pat = 1;
 	conf->he_6ghz_tx_ant_pat = 1;
 	conf->he_6ghz_reg_pwr_type = HE_REG_INFO_6GHZ_AP_TYPE_VLP;
+	conf->reg_def_cli_eirp_psd = -1;
+	conf->reg_sub_cli_eirp_psd = -1;
 	conf->he_oper_max_co_hosted_bssid_ind = 5;
 	conf->he_oper_cohosted_bss = 1;
 	conf->bss_color_change_timeout  = BSS_COLOR_CHANGE_TIMEOUT_DEFAULT;
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 0f2c14c21..84fb9779d 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1392,6 +1392,10 @@ struct hostapd_config {
 	u8 he_6ghz_rx_ant_pat;
 	u8 he_6ghz_tx_ant_pat;
 	u8 he_6ghz_reg_pwr_type;
+
+	int reg_def_cli_eirp_psd;
+	int reg_sub_cli_eirp_psd;
+
 	bool require_he;
 
 	u32 bss_color_change_timeout; /* in seconds */
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index a787ec843..ab1157e39 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7715,7 +7715,11 @@ u8 * hostapd_eid_txpower_envelope(struct hostapd_data *hapd, u8 *eid)
 		tx_pwr_intrpn = REGULATORY_CLIENT_EIRP_PSD;
 
 		/* Default Transmit Power Envelope for Global Operating Class */
-		tx_pwr = REG_PSD_MAX_TXPOWER_FOR_DEFAULT_CLIENT * 2;
+		if (hapd->iconf->reg_def_cli_eirp_psd != -1)
+			tx_pwr = hapd->iconf->reg_def_cli_eirp_psd;
+		else
+			tx_pwr = REG_PSD_MAX_TXPOWER_FOR_DEFAULT_CLIENT * 2;
+
 		eid = hostapd_add_tpe_info(eid, tx_pwr_count, tx_pwr_intrpn,
 					   REG_DEFAULT_CLIENT, tx_pwr);
 
@@ -7724,7 +7728,10 @@ u8 * hostapd_eid_txpower_envelope(struct hostapd_data *hapd, u8 *eid)
 		if (iconf->he_6ghz_reg_pwr_type ==
 		    HE_REG_INFO_6GHZ_AP_TYPE_INDOOR) {
 			/* TODO: Extract PSD limits from channel data */
-			tx_pwr = REG_PSD_MAX_TXPOWER_FOR_SUBORDINATE_CLIENT * 2;
+			if (hapd->iconf->reg_sub_cli_eirp_psd != -1)
+				tx_pwr = hapd->iconf->reg_sub_cli_eirp_psd;
+			else
+				tx_pwr = REG_PSD_MAX_TXPOWER_FOR_SUBORDINATE_CLIENT * 2;
 			eid = hostapd_add_tpe_info(eid, tx_pwr_count,
 						   tx_pwr_intrpn,
 						   REG_SUBORDINATE_CLIENT,
-- 
2.43.0

